src/sorting/countingsort.lua

```diff
@@ -0,0 +1,58 @@
+function counting_sort(arr, key_function)
+     -- If no key_function is provided, use identity function
+     key_function = key_function or function(x) return x end
+
+     -- Step 1: Find the range of keys (min_key and max_key)
+     local min_key, max_key = math.huge, -math.huge
+     for i = 1, #arr do
+         local key = key_function(arr[i])
+         if key < min_key then
+             min_key = key
+         end
+         if key > max_key then
+             max_key = key
+         end
+     end
+
+     -- Step 2: Initialize the count array
+     local count = {}
+     for i = min_key, max_key do
+         count[i] = 0
+     end
+
+      -- Step 3: Count the occurrences of each key. In this case key is same as arr[i]
```

Comment:
Unnecessary comment, also slightly misleading (keys are produced by the key function).